### PR TITLE
Fix DH/DHX named groups to not overwrite the private key length.

### PR DESCRIPTION
**Pull request by slontis**
The only reason(s) the DH private key length should be set are:
(1) The user sets it during key generation via EVP_PKEY_CTX_set_params
    using OSSL_PKEY_PARAM_DH_PRIV_LEN.
(2) When loading a PKCS3 (DH) key the optional value
    'privateValueLength' is set.

Now that the named groups contain a value for 'q' there is no reason to
automatically overwrite the private key length.

Issue detected by @davidmakepeace

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/dh/dh_group_params.c [hunk 1]
```diff
@@ -76,7 +75,6 @@ void ossl_dh_cache_named_group(DH *dh)
             dh->params.q = (BIGNUM *)ossl_ffc_named_group_get_q(group);
         /* cache the nid */
         dh->params.nid = ossl_ffc_named_group_get_uid(group);
-        dh->length = BN_num_bits(dh->params.q);
```
Comment (t8m) [inline crypto/dh/dh_group_params.c:79]: What about keeping it with `if (dh->length == 0)`?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/dh/dh_group_params.c [hunk 2]
```diff
@@ -34,7 +34,6 @@ static DH *dh_param_init(OSSL_LIB_CTX *libctx, const DH_NAMED_GROUP *group)
 
     ossl_ffc_named_group_set_pqg(&dh->params, group);
     dh->params.nid = ossl_ffc_named_group_get_uid(group);
-    dh->length = BN_num_bits(dh->params.q);
```
[the remark this answers is not in the document]
Comment (slontis) [reply crypto/dh/dh_group_params.c:37]: It looks wrong in 1.1.1.

File: crypto/dh/dh_group_params.c (hunk 1 above)
Comment (slontis) [reply crypto/dh/dh_group_params.c:79]: There is no point. The gen code will use BN_num_bits(dh->params.q) if the length is not set. This is the main reason for setting it (i.e. to set it to something different from the default).

**Review by t8m — APPROVED** [review]
The documentation is:

```
       The DH_get_length() and DH_set_length() functions get and set the
       optional length parameter associated with this DH object. If the length
       is nonzero then it is used, otherwise it is ignored. The length
       parameter indicates the length of the secret exponent (private key) in
       bits.
```

That makes me convinced this is correct change.

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**